### PR TITLE
Add 802.11r fast transition support to wifi.lua

### PR DESCRIPTION
**Pull request by muddyfeet**
Adds support for 802.11r fast transition to luci.  

The R0KH & R1KH options are not perfect as it would be better to separate out each option and do some sanity checking on the values, but that is beyond my capabilities.

Also addresses the issue of #844

### CODE REVIEW

**Comment by hnyman** [comment]
Thanks, I will likely take your commit, test it and polish it a bit, add a few value checks etc.

Some detail comments that you might fix before that:
* You forgot to sign-off your code. Could you please amend the commit and sign it off properly with your name etc.  (if you need advice on squashing & updating the PR, it can be found in https://github.com/openwrt/luci/blob/master/CONTRIBUTING.md#advice-on-pull-requests )
* The commit title should clearly say that it adds 802.11r support. "update wifi.lua" says nothing about the purpose.
* You look for the hostapd executable, but the result of that lookup it not used anyhow, as far as I see at the first glance.

(Or if you want to be lazy, I can "adopt" the code, edit it and commit it later signed off by me)

**Comment by muddyfeet** [comment]
My purpose of checking hostapd was to see if we were using wpad-mini or the full wpad/hostap.  I am assuming that there is no support for 802.11r in wpad-mini.

I'll update the pull request.

**Comment by hnyman** [comment]
> My purpose of checking hostapd was to see if we were using wpad-mini or the full wpad/hostap. 

Yes, the check itself is ok, but you do not use /check the value of that check later.

This is an example from the same file, where the result of that check is used in if statement.
https://github.com/muddyfeet/luci/blob/a83068af59630004cdfc4a6c1c1d1f2ac8615be3/modules/luci-mod-admin-full/luasrc/model/cbi/admin_network/wifi.lua#L1110

**Comment by muddyfeet** [comment]
Ah yes, I see what you mean.  That line can go - it's not needed here.

Will update.

**Comment by muddyfeet** [comment]
@hnyman not sure how to merge the 3 commits into one.  I'm editing the file online (not using a local fork)

**Comment by hnyman** [comment]
In that case, just leave them there. I will squash them.

**Comment by muddyfeet** [comment]
Thanks

**Comment by hnyman** [comment]
Squashed and committed manually as https://github.com/openwrt/luci/commit/171c06146ad2a45b1026e3a9779766855964fc2d

 (plus my small additions as https://github.com/openwrt/luci/commit/02bed261881dd60f212f81e7ed3e3afeb4bd890d )

This also closes #844

**Comment by hnyman** [comment]
@jow- 
Any hints how this kind of option values could be nicely split, edited and validated in LuCI?

I don't remember seeing quite this earlier.

The challenging 80211r uci options are comma-separated combinations of three items:
r0kh = MAC,string,hexstring
r1kh = MAC,MAC,hexstring

```
        list r0kh '74:44:00:CC:AA:E7,338899001122,44220000abcdefabcdefabcdefabcdef'
        list r1kh 'DC:EF:09:BB:FF:E7,77:88:88:00:00:22,44220000abcdefabcdefabcdefabcdef'
        list r1kh 'DC:33:44:BB:FF:E7,33:33:88:00:00:22,44220000abcdefabcdefabcdefabcdef'
```

Currently they are just text fields:
image

I think that the datatype validation function for the combined option could be built (rather easily?), but is there a nice way to split a field to three fields for editing in the GUI and then saving as a combined field string? I think that it would require actual coding, as CBI is probably not flexible enough.

File: modules/luci-mod-admin-full/luasrc/model/cbi/admin_network/wifi.lua [hunk 1]
```diff
@@ -900,18 +900,74 @@ end
...
+
+	ieee80211r = s:taboption("encryption", Flag, "ieee80211r", translate("802.11r Fast Transition"))
+	ieee80211r:depends({mode="ap", encryption="wpa"})
+	ieee80211r:depends({mode="ap", encryption="wpa2"})
+	ieee80211r:depends({mode="ap-wds", encryption="wpa"})
+	ieee80211r:depends({mode="ap-wds", encryption="wpa2"})
+	if has_ap_eap then
+		ieee80211r:depends({mode="ap", encryption="psk"})
+		ieee80211r:depends({mode="ap", encryption="psk2"})
+	end
```
Comment (simonsmh) [inline modules/luci-mod-admin-full/luasrc/model/cbi/admin_network/wifi.lua]: Is ap-wds under psk/psk2 working in 802.11r?

**Pull request closed without merging**